D3 simple bar charts

damnko opened the pull request:
* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
I have added a couple of bar charts created using D3.js: a simple bar chart and a grouped bar chart.

* **What is the current behavior?** (You can also link to an open issue here)
The grouped bar chart is not responsive, the rest seems to work.

* **Other information**:
I was actually wondering why should I use D3 to create simple/standard charts since there are many other simpler libraries to achieve the same. For this reason I am quite sure that you won't need what I have done but I have created a couple of them anyway to get an initial feedback on they way I handled it.
I found it difficult to make the charts responsive.

nnixaa commented on src/app/pages/charts/components/D3Js/D3Js.component.ts:
```diff
@@ -0,0 +1,25 @@
+    providers: []
+})
+
+export class D3Js implements OnInit{
```
`OnInit{` => `OnInit {`

nnixaa replied:
Overall - please use a space before the {

nnixaa commented on src/app/pages/charts/components/D3Js/D3Js.component.ts:
```diff
@@ -0,0 +1,25 @@
+export class D3Js implements OnInit{
+    data:any;
+
+    constructor( private _d3JsService:D3JsService){
```
`constructor( private` => `constructor(private`

nnixaa commented on src/app/pages/charts/components/D3Js/D3Js.component.ts:
```diff
@@ -0,0 +1,25 @@
+    constructor( private _d3JsService:D3JsService){
+    }
+
+    ngOnInit(){
```
same here

nnixaa commented on src/app/theme/components/baD3Chart/baD3Chart.service.ts:
```diff
@@ -0,0 +1,253 @@
+        
+        this.xAxis = d3.axisBottom(this.xScale);
+        this.yAxis = d3.axisLeft(this.yScale)
+                        .ticks(10, this.chartSetup.data.yAxisFormat ? this.chartSetup.data.yAxisFormat : "");
```
Please use single quotes everywhere.

nnixaa commented on src/app/theme/components/baD3Chart/baD3Chart.component.ts:
```diff
@@ -0,0 +1,63 @@
+    }
+
+    setup(){
+        this._chartService.chartSetup = {
```
I would say it's better to pass this configuration directly into the draw method of a specific chart type as in this case we won't save any state inside the service making it more-or-less stateless.

nnixaa commented on src/app/theme/components/baD3Chart/baD3Chart.service.ts:
```diff
@@ -0,0 +1,253 @@
+    private x1Scale;
+    private yScale;
+    private xAxis;
+    private yAxis;
```
Same for all service variables - do we really need to store them as it adds another complexity level as we need to think of them as in any place in the service we can't be sure what they are now, as they are available in all the service code?

damnko commented:
Hi I've made the code changes you requested, I hope I haven't missed anything.

The only thing I was wondering is whether should I put `BaD3ChartService` in the `baD3Chart.component` providers or should I use it as a single instance on `nga.module`. For the moment I left it in `baD2Chart.component` but I would have preferred to do the latter, but I didn't know whether to export the `BaD3ChartService` class in a `index.ts` file or not and then import it in `nga.module`.

I will wait for your feedback, thanks.

damnko commented:
Hi, I have made small changes to the code style:
- converted _4 spaces_ to _2 spaces_ in both `.ts` and `.html` files
- added `;` where needed
- other minor spacing _issues_

nnixaa commented:
Hey @damnko, sorry for taking this long. Checked again and all looks great! Could you please rebase master on your code so that I can merge it? Thanks.

damnko commented:
Hi, thanks for getting and no prob about the delay.
Since it's the first time that i _rebase_ should I push _master_ after rebasing and create a new PR? Or should I do something like _pull_ the latest _master_ and then applying the changes to my _d3-charts_ branch and the update this PR?
Hope you can help and thanks for your understanding.

nnixaa commented:
@damnko no problem, just pull the latest from master, rebase master onto your branch and update the PR.

damnko commented:
Ok hopefully I made it, please let me know if something is wrong so I can act accordingly. Thanks

ESadouski commented:
Hey @damnko 
Sorry for delay, but could you recreate it?

damnko commented:
Hi, what do you mean by _recreate it_?

lugovsky commented:
@damnko For some reason in this PR there are commits by @nnixaa and @ESadouski which is not quite right. It seems like something was broken after rebase...
It would be more safe if you create new branch from current master and `cherry-pick`ed all your commits there. It shouldn't take much time I think.

damnko commented:
Ok sure no problem I will do it asap.

phamdt commented:
anything happening with this?
